charts/core/src/Chart/hooks/useChart.ts

```diff
@@ -41,17 +41,99 @@ echarts.use([
-  onZoomSelect,
-}: ChartHookProps) {
+export function useChart({ theme, onChartReady, zoomSelect }: ChartHookProps) {
   const chartRef = useRef(null);
   const chartInstanceRef = useRef<echarts.EChartsType | undefined>();
   const [chartOptions, setChartOptions] = useState(
     getDefaultChartOptions(theme),
   );
 
+  function handleDataZoom(params: any) {
+    if (zoomSelect?.onZoomSelect) {
+      if (zoomSelect?.xAxis && zoomSelect?.yAxis) {
+        if (params.batch) {
+          const { startValue: xStart, endValue: xEnd } = params.batch[0];
+          const { startValue: yStart, endValue: yEnd } = params.batch[1];
+          zoomSelect.onZoomSelect({
+            xAxis: { startValue: xStart, endValue: xEnd },
+            yAxis: { startValue: yStart, endValue: yEnd },
+          });
+        }
+      } else if (zoomSelect?.xAxis || zoomSelect?.yAxis) {
+        const axis = zoomSelect?.xAxis ? 'xAxis' : 'yAxis';
+        const zoomEventResponse: ZoomSelectionEvent = {};
+
+        if (params.startValue && params.endValue) {
+          zoomEventResponse[axis] = {
+            startValue: params.startValue,
+            endValue: params.endValue,
+          };
+          zoomSelect.onZoomSelect(zoomEventResponse);
+        } else if (params.batch) {
+          const { startValue, endValue } = params.batch[0];
+          zoomEventResponse[axis] = {
+            startValue,
+            endValue,
+          };
+          zoomSelect.onZoomSelect(zoomEventResponse);
+        }
+      }
+    }
+
+    /**
+     * If start is not 0% or end is not 100%, that means that the 'dataZoom'
+     * event was triggered by an actual zoom. Since we don't want to actually
+     * zoom on the current data, but rather provide the new values to the passed
+     * in handler, we dispatch an action to essentially override the zoom.
+     */
+    const isZoomed = params?.start !== 0 || params?.end !== 100;
+
+    if (chartInstanceRef.current && isZoomed) {
+      chartInstanceRef.current.dispatchAction({
+        type: 'dataZoom',
+        start: 0, // percentage of starting position
+        end: 100, // percentage of ending position
+      });
+    }
+  }
+
+  /**
+   * Meant to be called once on initial render
+   */
+  function onInitialRender() {
+    if (!chartInstanceRef.current) {
+      return;
+    }
+
+    const chartInstance = chartInstanceRef.current;
+
+    if (zoomSelect?.xAxis || zoomSelect?.yAxis) {
+      /**
+       * Zooming is built into echart via the toolbar. By default, a user
+       * has to click the "dataZoom" button to enable zooming. We however hide
+       * this button and want it turned on by default. This is done by dispatching
+       * an action to enable the "dataZoomSelect" feature, as if it were clicked.
+       */
+      chartInstance.dispatchAction({
+        type: 'takeGlobalCursor',
+        key: 'dataZoomSelect',
+        dataZoomSelectActive: true,
+      });
+
+      chartInstance.on('dataZoom', handleDataZoom);
```

Comment:
If I understand correctly, even though `onInitialRender` is only called once on initial render, `handleDataZoom` will always be called when someone zooms?

Reply:
Correct. We remove `onInitialRender` as a render callback, but we never remove `handleDataZoom` as a dataZoom callback.